### PR TITLE
make pipeline to run benchmarking analysis

### PR DESCRIPTION
**Pull request by jishuxu**
In this PR,
1) build a pipeline to run benchmarking analysis automatically. 
2) remove common functions from `picard_QC` analysis and `confounding_factor` analysis R scripts
3) update wdl style
4) build a benchmarking docker image(not analysis-tools docker). This docker is stored in private bucket and include all the benchmarking analysis scripts(*.R and *.py)
 5) test results are here `https://job-manager.mint-dev.broadinstitute.org/jobs/b23b902e-258f-415e-8d14-66bd31f9badb`

### CODE REVIEW

File: benchmarking/smartseq2/combine_QCs/run_combine_QCs.wdl [hunk 1]
```diff
@@ -7,21 +7,20 @@ task CombineQCs{
...
-    gsutil cp ${src_dir}/combine_QC_quant_metrics.R ./
-    Rscript combine_QC_quant_metrics.R  --datafile ${datafile} --metrics ${metfile} --nthreshold ${nthreshold} --gtf ${gtf_file} --out ${output_name}
+    Rscript /usr/local/scripts/combine_QC_quant_metrics.R  --datafile ${datafile} --metrics ${metfile} --nthreshold ${nthreshold} --gtf ${gtf_file} --out ${output_name}
   }
   output {
     File combined_results = "${output_name}_metrics_combined.csv"
   }
   runtime {
-    docker:"gcr.io/broad-dsde-mint-dev/analysis-tools:0.0.5"
+    docker:"gcr.io/broad-dsde-mint-dev/benchmarking-tools:0.0.1"
```
Comment (rexwangcc) [inline benchmarking/smartseq2/combine_QCs/run_combine_QCs.wdl]: What kind of tools are included in this new docker image?

Comment (rexwangcc) [reply benchmarking/smartseq2/combine_QCs/run_combine_QCs.wdl]: NVM, I found the dockerfile below.

File: docker/benchmarking-tools/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+RUN Rscript -e "install.packages('rsvd')"
+RUN Rscript -e "install.packages('factoextra')"
+RUN Rscript -e "install.packages('fpc')"
+RUN Rscript -e "install.packages('NbClust')"
+
+# Add benchmarking scripts
+WORKDIR /usr/local/scripts/
+ENV PATH /usr/local/scripts/:$PATH
+COPY ./*.R /usr/local/scripts/
+COPY ./*.py /usr/local/scripts/
```
[the remark this answers is not in the document]
Comment (jishuxu) [reply docker/benchmarking-tools/Dockerfile]: `benchmarking-tools` docker will be as it is and `analysis-tools` docker can be updated more often

[the remark this answers is not in the document]
Comment (rexwangcc) [reply docker/benchmarking-tools/Dockerfile]: I see. That "freeze" makes sense to me.

File: benchmarking/smartseq2/R/analysis_functions.R [hunk 3]
```diff
@@ -361,3 +361,193 @@ RunSNNCluster <- function(mat.log) {
...
+  ## combine wiht QC
+  mlist <- match(colnames(met.core), names(cnt.ratio))
+  x <- data.frame(cnt.ratio[mlist])
+  colnames(x) <- 'detected_ratio'
+  mlist <- match(colnames(met.core), names(mt.ratio))
+  y <- data.frame(mt.ratio[mlist])
+  colnames(y) <- 'MT_ratio'
+  met.core <- rbind(t(x), t(y), met.core)
+  return(met.core)
+}
```
Comment (rexwangcc) [inline benchmarking/smartseq2/R/analysis_functions.R]: nit: new line at the end (optional)

**Comment by rexwangcc** [comment]
LGTM, dropped some minor comments. Also I have similar comments with @ambrosejcarr on the figures.

**Comment by ambrosejcarr** [comment]
I would love to see a `README.md` file in the benchmarking folder that links out to the rendered example output of the scripts that you showed me via slack. I feel like we're adding a lot of code here and it won't be clear to a user unless we add a bit more documentation.

File: benchmarking/smartseq2/RunBenchmarkingAnalysis.wdl [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+workflow RunBenchmarkingAnalysis {
+  String scripts_dir
+  String base_datafile
+  String updated_datafile
+  String output_name
+  String gtf_file
+  String metadata_file
+  String metKeys
+  String base_metrics
+  String updated_metrics
```
Comment (mckinsel) [inline benchmarking/smartseq2/RunBenchmarkingAnalysis.wdl]: Could you double-check these types? I think you have a lot of files typed as `String`.

**Comment by kbergin** [comment]
@jishuxu @ambrosejcarr I see the readme was updated to have a link to each benchmarking html report. If there could be maybe a title and a description of what the benchmarking is, what data was used, what pipelines were compared. And if the main skylab readme could be updated, at the bottom where it says "TODO Onboarding ..." can we replace that with Benchmarking Analyses and a description of what folders to go to for those analyses? Later we can replace that with a link out to Unity! :)

File: pipelines/smartseq2_single_sample/configuration.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
```
Comment (dshiga) [inline pipelines/smartseq2_single_sample/configuration.json]: I would call this options.json since that's what cromwell calls it and that's what we call it elsewhere.

File: benchmarking/smartseq2/RunBenchmarkingAnalysis.wdl [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  String output_name
+  String gtf_file
+  String metadata_file
+  String metadata_keys
+  String metKeys
+  String base_metrics
+  String updated_metrics
+  Int    npcs
+
+  meta {
```
Comment (dshiga) [inline benchmarking/smartseq2/RunBenchmarkingAnalysis.wdl]: meta should be above the variable declarations on line 4 and have an empty line underneath

File: benchmarking/smartseq2/RunBenchmarkingAnalysis.wdl [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  String base_datafile
+  String updated_datafile
+  String output_name
+  String gtf_file
+  String metadata_file
+  String metadata_keys
+  String metKeys
+  String base_metrics
+  String updated_metrics
+  Int    npcs
```
Comment (dshiga) [inline benchmarking/smartseq2/RunBenchmarkingAnalysis.wdl]: Extra spaces should be removed here

File: benchmarking/smartseq2/RunBenchmarkingAnalysis.wdl [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+  meta {
+    description: "Process SmartSeq2 Benchmarking pipeline"
```
Comment (dshiga) [inline benchmarking/smartseq2/RunBenchmarkingAnalysis.wdl]: I could use more information in this description. Could you summarize what the workflow does in a little more detail?

**Comment by dshiga** [comment]
Still reviewing, but I think more documentation is needed, partly in the WDL and partly in the README, to explain what all this code is in this directory, what it does and how it's intended to be used.

**Comment by dshiga** [comment]
I think it would be nice to check in the test WDL in this PR too, since it is closely related to the benchmarking WDL and in practice we'll run them together a lot.

File: benchmarking/smartseq2/RunBenchmarkingAnalysis.wdl [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  meta {
+    description: "Run SmartSeq2 Benchmarking pipeline. 4 modules are included in this pipeline.QC test, Comparative tests, Reproducibility test and Confounding factors test."
+  }
+  String base_datafile
+  String updated_datafile
+  String output_name
+  String gtf_file
+  String metadata_file
+  String metadata_keys
+  String metKeys
```
Comment (dshiga) [inline benchmarking/smartseq2/RunBenchmarkingAnalysis.wdl]: Should be met_keys

File: benchmarking/smartseq2/BenchmarkingTasks.wdl [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+  command {
+    set -e
+    cp  /usr/local/scripts/*.R ./
+    Rscript -e 'rmarkdown::render("Compare_data_matrix.R", output_file="${output_name}_Compare_data_matrix.html")' --args \
+      "--matrix1=${base_datafile} --matrix2=${updated_datafile} --metaKeys=${metadata_keys} --metadata_file=${metadata_file} --output_prefix=${output_name}"
+  }
+  output {
+    File html = "${output_name}_Compare_data_matrix.html"
+    File memnbership = "${output_name}_tsne_SNN_cluster_data_matrix_comparison.csv"
```
Comment (dshiga) [inline benchmarking/smartseq2/BenchmarkingTasks.wdl]: typo, should be membership

File: benchmarking/smartseq2/BenchmarkingTasks.wdl [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+  }
+  runtime {
+    docker:"gcr.io/broad-dsde-mint-dev/benchmarking-tools:0.0.1"
+    memory:"7.5 GB"
+    disks: "local-disk 50 HDD"
+    preemptible: 5
+  }
+}
```
Comment (dshiga) [inline benchmarking/smartseq2/BenchmarkingTasks.wdl]: new line between tasks please

File: benchmarking/smartseq2/BenchmarkingTasks.wdl [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    cp  /usr/local/scripts/*.R ./
+    Rscript -e 'rmarkdown::render("Compare_data_matrix.R", output_file="${output_name}_Compare_data_matrix.html")' --args \
+      "--matrix1=${base_datafile} --matrix2=${updated_datafile} --metaKeys=${metadata_keys} --metadata_file=${metadata_file} --output_prefix=${output_name}"
+  }
+  output {
+    File html = "${output_name}_Compare_data_matrix.html"
+    File memnbership = "${output_name}_tsne_SNN_cluster_data_matrix_comparison.csv"
+  }
+  runtime {
+    docker:"gcr.io/broad-dsde-mint-dev/benchmarking-tools:0.0.1"
```
Comment (dshiga) [inline benchmarking/smartseq2/BenchmarkingTasks.wdl]: space after `:`

File: benchmarking/smartseq2/BenchmarkingTasks.wdl [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+  } 
+
+}
+task RunConfoundingFactorAnalysis{
+  File base_metrics
+  File updated_metrics
+  File metadata_file
+  File base_datafile
+  File updated_datafile
+  String metaKeys
```
Comment (dshiga) [inline benchmarking/smartseq2/BenchmarkingTasks.wdl]: meta_keys

File: benchmarking/smartseq2/RunBenchmarkingAnalysis.json [hunk 14]
```diff
@@ -0,0 +1,13 @@
+{
+  "RunBenchmarkingAnalysis.scripts_dir": "gs://broad-dsde-mint-dev-teststorage/pipeline_testing_scripts",
+  "RunBenchmarkingAnalysis.base_datafile": "gs://broad-dsde-mint-dev-teststorage/benchmarking/ss2/STAR_rsem_merged_tpm.csv'",
+  "RunBenchmarkingAnalysis.gtf_file": "gs://hca-dcp-mint-test-data/reference/GRCh38_Gencode/gencode.v27.primary_assembly.annotation.gtf",
+  "RunBenchmarkingAnalysis.metadata_file":"gs://broad-dsde-mint-dev-teststorage/sloan_sa_astrocytes/Sra_meta_groups_iPSC_730.csv",
+	"RunBenchmarkingAnalysis.updated_datafile": "gs://broad-dsde-mint-dev-teststorage/benchmarking/ss2/call-MergeTPM/shard-0/HISAT2_rsem_merged_tpm.csv",
```
Comment (dshiga) [inline benchmarking/smartseq2/RunBenchmarkingAnalysis.json]: should get rid of tabs here and below

File: benchmarking/smartseq2/RunBenchmarkingAnalysis.json [hunk 15]
```diff
@@ -0,0 +1,13 @@
+{
+  "RunBenchmarkingAnalysis.scripts_dir": "gs://broad-dsde-mint-dev-teststorage/pipeline_testing_scripts",
+  "RunBenchmarkingAnalysis.base_datafile": "gs://broad-dsde-mint-dev-teststorage/benchmarking/ss2/STAR_rsem_merged_tpm.csv'",
+  "RunBenchmarkingAnalysis.gtf_file": "gs://hca-dcp-mint-test-data/reference/GRCh38_Gencode/gencode.v27.primary_assembly.annotation.gtf",
+  "RunBenchmarkingAnalysis.metadata_file":"gs://broad-dsde-mint-dev-teststorage/sloan_sa_astrocytes/Sra_meta_groups_iPSC_730.csv",
```
Comment (dshiga) [inline benchmarking/smartseq2/RunBenchmarkingAnalysis.json]: space after `:`

File: docker/benchmarking-tools/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+    google-cloud-storage \
+    git+git://github.com/HumanCellAtlas/pipeline-tools.git
+
+# Fix cannot import name 'opentype' error
+RUN pip3 install --upgrade google-auth-oauthlib
+# Mount Google Cloud secret key
+ADD broad-dsde-mint-dev-analysis-tools.json /usr/secrets/broad-dsde-mint-dev-analysis-tools.json
+
+# Set the environment variable for Python
+ENV GOOGLE_APPLICATION_CREDENTIALS=/usr/secrets/broad-dsde-mint-dev-analysis-tools.json
```
Comment (dshiga) [inline docker/benchmarking-tools/Dockerfile]: Are you downloading files from a bucket inside R? It would be preferable to pass those file paths to the task and let Cromwell / Pipelines API localize them for you. That way you wouldn't need to bake credentials into this docker image.

File: benchmarking/smartseq2/merge_picard_metrics/merge_picard_metrics.wdl [hunk 17]
```diff
@@ -5,14 +5,13 @@ task MergePicardMetrics {
   String output_name
   command {
     set -e
-    gsutil cp  ${target_dir}/merge_picard_mets.py ./   
-    python merge_picard_mets.py  -u ${uuid} -m ${met_name} -o ${output_name}
+    python /usr/local/scripts/merge_picard_mets.py  -u ${uuid} -m ${met_name} -o ${output_name}
```
Comment (dshiga) [inline benchmarking/smartseq2/merge_picard_metrics/merge_picard_metrics.wdl]: target_dir above isn't used, can be removed

Comment (dshiga) [reply benchmarking/smartseq2/merge_picard_metrics/merge_picard_metrics.wdl]: Would be preferable to pass in an Array[File] to this task so it can just work with files locally rather than having to download and parse the metadata and then download files from the bucket in Python.

You would need to define that Array[File] in your inputs json file, but you could do that in a couple of ways:

1. Declare the metrics file you need as workflow outputs in the test workflow. When the test workflow finishes, copy and paste the list of metrics files from the workflow metadata into your inputs json file.
2. It might be easier to have one big WDL that runs the test WDL and then runs the benchmarking WDL. That would avoid the problem of needing to copy files or copy/paste lists of files altogether. You mentioned some issues with that approach related to various samples failing, but we could work to make the WDL able to handle that better. (We can set Cromwell server config to have invalidate-bad-cache-results = true to fix one issue.)

File: pipelines/smartseq2_single_sample/configuration.json (hunk 5 above)
Comment (jishuxu) [reply pipelines/smartseq2_single_sample/configuration.json]: changed

File: benchmarking/smartseq2/RunBenchmarkingAnalysis.json [hunk 18]
```diff
@@ -0,0 +1,13 @@
+{
+  "RunBenchmarkingAnalysis.scripts_dir": "gs://broad-dsde-mint-dev-teststorage/pipeline_testing_scripts",
```
Comment (ambrosejcarr) [inline benchmarking/smartseq2/RunBenchmarkingAnalysis.json]: These files should be in the public mint bucket, I think. @dshiga @TimothyTickle Do you guys agree?

**Comment by jishuxu** [comment]
@ambrosejcarr  thanks

This PR has been hanging here for too long and the key components of it had been reviewed, approved and merged once . The Original ticket for this PR was just to write a wdl to launch benchmarking tests automatically but it turned out much more than I expected.

File: benchmarking/smartseq2/RunBenchmarkingAnalysis.json (hunk 18 above)
Comment (jishuxu) [reply benchmarking/smartseq2/RunBenchmarkingAnalysis.json]: we don't need this any more. all the scripts are in docker now. I will remove this line from json.
